Kernel/Bus/PCI/Device.cpp

```diff
@@ -61,4 +66,101 @@ void Device::disable_extended_message_signalled_interrupts()
     }
 }
 
+PCI::InterruptType Device::get_interrupt_type()
+{
+    return m_interrupts.m_type;
+}
+
+// Reserve `numbers_of_irqs` for this device. Returns the interrupt type
+// that was reserved. It is a noop for pin based interrupts as there
+// is nothing left to do. The second parameter `msi` is used by the
+// driver to indicate its intent to use message signalled interrupts.
+// MSI(x) is preferred over MSI if the device supports both.
+ErrorOr<InterruptType> Device::reserve_irqs(u8 number_of_irqs, bool msi)
+{
+    // Let us not allow partial allocation of IRQs for MSIx.
+    if (msi && is_msix_capable()) {
+        m_interrupts.m_start_irq = TRY(reserve_interrupt_handlers(number_of_irqs));
+        m_interrupts.m_irq_count = number_of_irqs;
+        m_interrupts.m_type = InterruptType::MSIX;
+        // If MSIx is available, disable the pin based interrupts
+        disable_pin_based_interrupts();
+        enable_extended_message_signalled_interrupts();
+    } else if (msi && is_msi_capable()) {
+        TODO();
+    }
+    return m_interrupts.m_type;
+}
+
+PhysicalAddress Device::msix_table_entry_address(u8 irq)
+{
+    auto index = irq - m_interrupts.m_start_irq;
```

Comment:
You're subtracting a `u8` from another `u8`, wouldn't this wrap around and guarantee `VERIFY(index >= 0)` below? Maybe cast this to an `int` first.